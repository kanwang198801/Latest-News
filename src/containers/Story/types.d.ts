interface StoriesState {

}
type ContainerState = StoriesState;

export { ContainerState };