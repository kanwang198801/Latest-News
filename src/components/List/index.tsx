import * as React from 'react';
import styles from './styles.module.css';
import { Link } from 'react-router-dom'
import ReactHtmlParser from 'react-html-parser';
function List({ items, link, type }) {
    let list = "";

    if (type === "story") {
        list = items.map((listItem: any) => (
            <div className={styles.listItem} key={`item - ${listItem.id}`}>
                <Link className="link" to={`${link}/${listItem.id}`}>
                    {listItem.title}
                </Link >
            </div>
        ))
    }
    else if (type === "comment") {
        list = items.map((listItem: any) => {
            const dateTime = new Date(listItem.time * 1000);
            const date = dateTime.getDate() + '/' + (dateTime.getMonth() + 1) + '/' + dateTime.getFullYear();
            return (
                <div className={styles.listItem} key={`item - ${listItem.id}`}>
                    <div>{ReactHtmlParser(listItem.text)}</div>
                    <h5><strong>By:</strong> {listItem.by} - {date}</h5>
                </div>
            );
        })
    }
    return (
        <div className={styles.list}>
            {list}
        </div>
    );
}
export default List;
