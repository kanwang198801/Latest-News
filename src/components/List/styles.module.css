.list {
  padding:15px;
  border:1px solid #ccc;
  background-color:#f5f5f5;
}
.listItem{
  margin-bottom:15px;
}