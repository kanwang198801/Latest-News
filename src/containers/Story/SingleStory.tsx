/*
 *
 * Single Story
 *
 */

import React, { useEffect, useState, memo } from 'react';
import { Helmet } from 'react-helmet';
import URL from '../../API';
import Theme from '../../components/Theme';
import { RouteComponentProps, Link } from 'react-router-dom';
import CommentList from '../../components/CommentList';
import Button from '../../components/Button';
import { StoryType, CommentType } from './types';

type SingleStoryParams = {
  id: string;
};
type Props = RouteComponentProps<SingleStoryParams>;

function SingleStory(props: Props) {
  const [loading, setLoading] = useState(true);
  const [commentloading, setCommentLoading] = useState(true);
  const [story, setStory] = useState<StoryType>({
    by: '',
    descendants: 0,
    id: 0,
    kids: [],
    score: 0,
    time: 0,
    type: '',
    title: '',
    url: '',
  });
  const [comments, setComments] = useState<CommentType[]>([]);
  const loadStory = async () => {
    const singleProjectID = props.match.params.id;
    fetch(`${URL}/item/${singleProjectID}.json?print=pretty`)
      .then((res) => res.json())
      .then((response) => {
        if (response !== null) setStory(response);
        setLoading(false);
      })
      .catch((error) => console.log(error));
  };
  const loadComments = async (commentIDs: number[]) => {
    const topCommentIDs = commentIDs.slice(0, 20);
    const list: CommentType[] = await Promise.all(
      topCommentIDs.map((comentId) => {
        const singleComment = fetch(`${URL}/item/${comentId}.json?print=pretty`)
          .then((res) => res.json())
          .then((response) => {
            return response;
          })
          .catch((error) => console.log(error));
        return singleComment;
      })
    );
    setComments(list);
    setCommentLoading(false);
  };

  useEffect(() => {
    loadStory();
  }, []);

  useEffect(() => {
    if (story.type !== 'story') return;
    if (story.kids.length > 0) {
      loadComments(story.kids);
    }
  }, [story]);
  interface ListTypes {
    items: CommentType[];
  }
  const ListProps: ListTypes = {
    items: comments,
  };
  let date = '';
  if (story.time) {
    const dateTime = new Date(story.time * 1000);
    date = dateTime.getDate() + '/' + (dateTime.getMonth() + 1) + '/' + dateTime.getFullYear();
  }

  return (
    <Theme>
      <Helmet>
        <title>Story</title>
        <meta name="description" content="Story" />
      </Helmet>
      {!loading ? (
        story.type === 'story' ? (
          <>
            <Link to="/">
              <Button>Back</Button>
            </Link>
            <h1>{story.title}</h1>
            <h3>
              Arthur: {story.by} - {date}
            </h3>
            <a href={story.url} target="_black">
              View details
            </a>
            {!commentloading ? (
              comments.length > 0 && (
                <>
                  <h3>Comments</h3>
                  <CommentList {...ListProps} />
                </>
              )
            ) : (
              <p>loading comments...</p>
            )}
          </>
        ) : (
          <p>Not found</p>
        )
      ) : (
        <p>loading...</p>
      )}
    </Theme>
  );
}

export default memo(SingleStory);
