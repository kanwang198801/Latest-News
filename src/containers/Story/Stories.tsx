/*
 *
 * Stories Listing
 *
 */

import React, { useEffect, useState, memo } from 'react';
import { Link, RouteComponentProps } from 'react-router-dom'
import { Helmet } from "react-helmet";
import URL from "../../API";
import ListItem from "../../components/ListItem";
import { StoryType } from "./types"

interface OwnProps extends RouteComponentProps { }
type Props = OwnProps;

function Stories(props: Props) {
    const [loading, setLoading] = useState(true);
    const [stories, setStories] = useState<StoryType[]>([]);
    useEffect(() => {
        fetch(
            `${URL}/topstories.json?print=pretty`
        ).then(res => res.json())
            .then(response => {
                const topTenStoryIDs: [] = response.slice(0, 10);
                let storiesTopTen: StoryType[] = [];

                topTenStoryIDs.forEach(topTenStoryID => {
                    fetch(
                        `${URL}/item/${topTenStoryID}.json?print=pretty`
                    )
                        .then(res => res.json())
                        .then(response => {
                            storiesTopTen.push(response);
                        }).catch(error => console.log(error));
                });
                setStories(storiesTopTen);
                setLoading(false);
            }
            )
            .catch(error => console.log(error));

    }, []);

    console.log("Object.keys(stories).length", Object.keys(stories).length);
    console.log("stories[0]", stories[0]);
    console.log(stories);
    // console.log(stories);
    // console.log(stories[0]);
    // console.log(stories.length);
    // console.log(typeof stories);
    return (
        <div>
            <Helmet>
                <title>Stories Listing</title>
                <meta name="description" content="Stories Listing" />
            </Helmet>
            {!loading ?
                Object.keys(stories).length !== 0 ? (<div>No data</div>) :
                    (<div>
                        {stories.map((storyItem) => (
                            <Link to={`/story/${storyItem.id}`} key={`item - ${storyItem.id}`}>
                                <ListItem storyItem />
                            </Link >
                        ))}
                    </div>
                    ) :
                <div>Loading...</div>}

        </div>
    );
}

export default memo(Stories);