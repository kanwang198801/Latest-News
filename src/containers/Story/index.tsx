/**
 *
 * Stories
 *
 */

import Stories from './Stories';
import SingleStory from './SingleStory';

export { Stories, SingleStory };
