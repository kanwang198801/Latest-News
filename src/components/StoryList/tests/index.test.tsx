/**
 * Testing our List component
 */

import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { StoryType, CommentType } from '../../../containers/Story/types';
import List from '../index';

interface storyListTypes {
  type: string;
  link: string;
  items: StoryType[];
}

describe('<StoryList >', () => {
  it('should render the List if loading was story list successful', () => {
    const props: storyListTypes = {
      type: 'story',
      link: 'https://google.com',
      items: [
        {
          by: 'pjc50',
          descendants: 22944177,
          id: 22943621,
          kids: [22944685, 22944096, 22944317],
          score: 95,
          title: '&quot;Permanent&quot; and &quot;free tier&quot;',
          time: 1587552301,
          type: 'comment',
          url: 'https://www.permanent.org/',
        },
        {
          by: 'pjc50',
          descendants: 22944174,
          id: 22943620,
          kids: [22944681, 22944092, 22944313],
          score: 95,
          title: '&quot;Permanent&quot; and &quot;free tier&quot;',
          time: 1587552300,
          type: 'comment',
          url: 'https://www.permanent.org/',
        },
      ],
    };
    const { container } = render(
      <Router>
        <List {...props} />
      </Router>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
