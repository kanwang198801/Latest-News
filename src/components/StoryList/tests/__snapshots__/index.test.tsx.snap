// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StoryList > should render the List if loading was story list successful 1`] = `
<div
  class="list"
>
  <div
    class="listItem"
  >
    <a
      class="link"
      href="/https://google.com/22943621"
    >
      &quot;Permanent&quot; and &quot;free tier&quot;
    </a>
  </div>
  <div
    class="listItem"
  >
    <a
      class="link"
      href="/https://google.com/22943620"
    >
      &quot;Permanent&quot; and &quot;free tier&quot;
    </a>
  </div>
</div>
`;
